rpcs3/rpcs3qt/cheat_manager.cpp

```diff
@@ -0,0 +1,840 @@
+	return true;
+}
+
+bool cheat_engine::resolve_script(u32& final_offset, const u32 offset, const std::string& red_script)
+{
+	enum operand
+	{
+		operand_equal,
+		operand_add,
+		operand_sub
+	};
+
+	auto do_operation = [](const operand op, u32& param1, const u32 param2) -> u32 {
+		switch (op)
+		{
+		case operand_equal: return param1 = param2;
+		case operand_add: return param1 += param2;
+		case operand_sub: return param1 -= param2;
+		}
+		ASSERT(false);
+	};
+
+	operand cur_op = operand_equal;
+	u32 index      = 0;
+
+	while (index < red_script.size())
+	{
+		if (red_script[index] >= '0' && red_script[index] <= '9')
+		{
+			std::string num_string;
+			for (; index < red_script.size(); index++)
+			{
+				if (red_script[index] < '0' || red_script[index] > '9')
+					break;
+
+				num_string += red_script[index];
+			}
+
+			u32 num_value = std::stoul(num_string);
+			do_operation(cur_op, final_offset, num_value);
+		}
+		else
+		{
+			switch (red_script[index])
+			{
+			case '$':
+			{
+				do_operation(cur_op, final_offset, offset);
+				index++;
+				break;
+			}
+			case '[':
+			{
+				// find corresponding ]
+				s32 found_close = 1;
+				std::string sub_script;
+				for (index++; index < red_script.size(); index++)
+				{
+					if (found_close == 0)
+						break;
+
+					if (red_script[index] == ']')
+						found_close--;
+					else if (red_script[index] == '[')
+						found_close++;
+
+					if (found_close != 0)
+						sub_script += red_script[index];
+				}
+
+				if (found_close)
+					return false;
+
+				// Resolves content of []
+				u32 res_addr = 0;
+				if (!resolve_script(res_addr, offset, sub_script))
+					return false;
+
+				// Tries to get value at resolved address
+				bool success;
+				u32 res_value = get_value<u32>(res_addr, success);
+
+				if (!success)
+					return false;
+
+				do_operation(cur_op, final_offset, res_value);
+			}
+			case '+':
+				cur_op = operand_add;
+				index++;
+				break;
+			case '-':
+				cur_op = operand_sub;
+				index++;
+				break;
+			case ' ': index++; break;
+			default: log_cheat.fatal("invalid character in redirection script"); return false;
+			}
+		}
+	}
+
+	return true;
+}
+
+template <typename T>
+std::vector<u32> cheat_engine::search(const T value, const std::vector<u32>& to_filter)
+{
+	std::vector<u32> results;
+
+	be_t<T> value_swapped = value;
+
+	if (Emu.IsStopped())
+		return {};
+
+	cpu_thread::suspend_all cpu_lock(nullptr);
+
+	if (to_filter.size())
+	{
+		for (const auto& off : to_filter)
+		{
+			if (vm::check_addr(off, 1, vm::page_info_t::page_allocated | vm::page_info_t::page_readable))
+			{
+				if (*vm::get_super_ptr<T>(off) == value_swapped)
+					results.push_back(off);
+			}
+		}
+	}
+	else
+	{
+		// Looks through mapped memory
+		for (u32 page_ind = (0x10000 / 4096); page_ind < (0xF0000000 / 4096); page_ind++)
+		{
+			if (vm::check_addr(page_ind * 4096, 1, vm::page_info_t::page_allocated | vm::page_info_t::page_readable))
```

Comment:
I know you are aware of this, leaving it here as a reminder.
```suggestion
			if (vm::check_addr(page_ind * 4096))
```